qt5.qtbase: fix cross

amjoseph-nixpkgs opened the pull request:
###### Description of changes

This commit fixes builds of `pkgsCross.*.qt5.qtbase` by adding the buildPlatform compiler to depsBuildBuild.  The `qtbase` build machinery expects to find it in the $PATH in unprefixed form.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Commit:
qt-5/modules/qtbase.nix: omit --host and --build configureFlags

QT's configure script doesn't understand these, so let's not pass them.

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

Commit:
qt5.qtbase: fix cross compilation

This commit fixes builds of `pkgsCross.*.qt5.qtbase` by:

- Adding the buildPlatform compiler to depsBuildBuild in qtbase.nix
  and qtModule.nix.  The `qtbase` build machinery expects to find it
  in the $PATH in unprefixed form.

- Setting the `PKG_CONFIG_SYSROOT_DIR` and `PKG_CONFIG_LIBDIR`
  environment variables when compiling a cross-targeted `qmake`.
  This is required; if these environment variables are unset,
  `qmake` won't even try to use `pkg-config`.

- Adding the `-device` and `-device-option` flags necessary for
  cross compilation to `configureFlags`.

- Adding the (one-entry at the moment) Rosetta Stone for QT-5 as a
  `let`-defined `qtPlatform` function which takes a nixpkgs platform
  and returns a QT-5 `mkspecs`-string.

Co-authored-by: Christoph Neidahl <christoph.neidahl@gmail.com>

amjoseph-nixpkgs commented:
This project will need to wait until after ZHF.

Mindavi commented:
Should go to staging

OPNA2608 commented:
The last force-push (aa35c5080517855e9bf633e190cebff8fbee1d99 to 6a79f70237d8aefec6e5836e29056f69f555bb1d) dropped the change to `qtbase`, it this still WIP or was that a mistake?

amjoseph-nixpkgs commented:
Thanks, I pushed the wrong branch.

It's definitely ready; I've been using it on my daily-driver laptop for months.

OPNA2608 reviewed (changes requested):
1. I think the commit title should refer to `libsForQt5.qtbase`
2. This still doesn't fix cross-compilation for me on its own, maybe the commit title should only clarify the part that it does fix (usage of a compiler for the buildPlatform)?
  - on the PR branch, it gets further in the build process but seems to require #220315 next
  - merged into staging via `nixpkgs-review`, some unrelated dependency-of-a-dependency fails

 …but I think it's a step in the right direction in general.

amjoseph-nixpkgs commented:
> 1. I think the commit title should refer to `libsForQt5.qtbase`

Done.

> * on the PR branch, it gets further in the build process but seems to require

Rebased on top of #220315.

> merged into staging via `nixpkgs-review`, some unrelated dependency-of-a-dependency fails

Fixed a merge conflict.

amjoseph-nixpkgs commented:
Rebased.

amjoseph-nixpkgs commented:
Last commit should help.  I have cups/postgres/gtk3-in-qt5 disabled on my system.

@ofborg build pkgsCross.aarch64-multiplatform.qt5.qtbase

amjoseph-nixpkgs commented:
@ofborg build pkgsCross.aarch64-multiplatform.qt5.qtbase

amjoseph-nixpkgs commented:
@OPNA2608 yep, it builds:

> pkgsCross.aarch64-multiplatform.qt5.qtbase on x86_64-linux — Success
[Details](https://github.com/NixOS/nixpkgs/pull/227900/checks?check_run_id=15975808058)

OPNA2608 reviewed (commented):
Can confirm that `qtbase` cross-compiles to armvl7-linux & aarch64-linux locally. Some questions/thoughts below.

amjoseph-nixpkgs commented on pkgs/development/libraries/qt-5/modules/qtbase.nix:
```diff
@@ -161,6 +171,13 @@ stdenv.mkDerivation (finalAttrs: {
 
     ./bin/syncqt.pl -version $version
+  '' + lib.optionalString (!stdenv.buildPlatform.canExecute stdenv.hostPlatform) ''
+    # QT's configure script will refuse to use pkg-config unless these two environment variables are set
+    export PKG_CONFIG_SYSROOT_DIR=/
+    export PKG_CONFIG_LIBDIR=${lib.getLib pkg-config}/lib
```
Well, the problem is that nixpkgs doesn't have a `qmake` package.  `qmake` is a build tool, so it properly ought to have a notion of targetPlatform, and it makes sense to talk about "cross-qmake".

Unfortunately we bundle qmake in with qtbase, which is a target library.  The result is a real headache.

I am not up to breaking out qmake as its own package; that's more work than I can take on (and would probably break other things, leading to people getting pissed off at me).

OPNA2608 replied:
Hmm, I don't think breaking out `qmake` is necessary for now… Maybe nice, but a whole load of extra effort and pain like you said.

I found [this](https://github.com/qt/qtbase/blob/42e4ae042a4c86e58bcb8b6d2d59ba4a988285b4/mkspecs/devices/common/linux_device_pre.conf#L31-L32). So QMake has the option to use a prefixed `pkg-config`, with the same prefix as the compiler & bintools, which seems to be exactly what we need? Patching it to `if(true)` makes QMake-dependent builds work _fine_\*.

Either we add `DISTRO_OPTS+=deb-multi-arch` to the list of QMake variables we echo into the generic config below (although this also turns on [something else](https://github.com/qt/qtbase/blob/e26a87f1ecc40bc8c6aa5b889fce67410a57a702/mkspecs/devices/common/linux_device_post.conf#L5-L9) which we maybe don't want, not sure of its impact?), or we patch only that bit to `if(true)`. Maybe even just echo in `QMAKE_PKG_CONFIG = $${CROSS_COMPILE}pkg-config` below to let it inherit the cross prefix when necessary. I think that would be less invasive than what I proposed earlier.

Except, even with that patched it still seems to require these two variables while building itself, so something might still be missing there… Feel free to consider this point resolved for now as I don't know how to address this better. But I'll prolly look into this stuff at some later point cus I would also like to have working Qt cross for Lomiri, which sadly includes some QMake projects.

\* <sub>Except that Qt module dependencies only get picked up by `qmake` when they're in `nativeBuildInputs`, both under cross and with `strictDeps` enabled. I also wanna look into this at some point…</sub>

amjoseph-nixpkgs replied:
So, I've sorted your suggestions from least-desirable to most-desirable:

> Either we add `DISTRO_OPTS+=deb-multi-arch` to the list of QMake variables we echo into the generic config below (although this also turns on [something else](https://github.com/qt/qtbase/blob/e26a87f1ecc40bc8c6aa5b889fce67410a57a702/mkspecs/devices/common/linux_device_post.conf#L5-L9) which we maybe don't want, not sure of its impact?)

Yeah not excited about the fact that it does random other stuff.

> or we patch only that bit to `if(true)`.

Nice, but then we have to deal with the patch possibly not applying in the future.  In general setting an environment variable should almost always be preferred over any kind of patch, if possible.

> Maybe even just echo in `QMAKE_PKG_CONFIG = $${CROSS_COMPILE}pkg-config` below to let it inherit the cross prefix when necessary.

I like this option.  Testing it now.

amjoseph-nixpkgs replied:
> I like this option. Testing it now.

Implemented in f81e02d9a771

amjoseph-nixpkgs commented on pkgs/development/libraries/qt-5/modules/qtbase.nix:
```diff
@@ -161,6 +171,13 @@ stdenv.mkDerivation (finalAttrs: {
+    # QT's configure script will refuse to use pkg-config unless these two environment variables are set
+    export PKG_CONFIG_SYSROOT_DIR=/
+    export PKG_CONFIG_LIBDIR=${lib.getLib pkg-config}/lib
+    echo "QMAKE_LFLAGS=''${LDFLAGS}" >> mkspecs/devices/${qtPlatformCross stdenv.hostPlatform}/qmake.conf
+    echo "QMAKE_CFLAGS=''${CFLAGS}" >> mkspecs/devices/${qtPlatformCross stdenv.hostPlatform}/qmake.conf
+    echo "QMAKE_CXXFLAGS=''${CXXFLAGS}" >> mkspecs/devices/${qtPlatformCross stdenv.hostPlatform}/qmake.conf
```
> Why do we fully override the flags QMake thinks it needs for the compilation, and why only when we're cross-compiling to a platform whose binaries we cannot execute locally?

Good point; that conditional ought to be `!=`.  Fixed in 487d5ad9ef3c.

> And at least in my attempts, these always end up empty.

Er, what ends up empty?

OPNA2608 replied:
> Er, what ends up empty?

The variables being set here.

```
[nix-shell:~/.cache/nixpkgs-review/pr-227900-5]$ nix-build --no-out-link ./nixpkgs/ -A libsForQt5.qtbase.dev --argstr crossSystem aarch64-linux
/nix/store/p1qhrrg12h28nknxsahyszlrrqagd0ni-qtbase-aarch64-unknown-linux-gnu-5.15.9-dev

[nix-shell:~/.cache/nixpkgs-review/pr-227900-5]$ cat /nix/store/p1qhrrg12h28nknxsahyszlrrqagd0ni-qtbase-aarch64-unknown-linux-gnu-5.15.9-dev/mkspecs/devices/linux-generic-g++/qmake.conf 
#
# Generic qmake configuration for building with g++ on generic devices.
#
# A minimal configure line could look something like this:
# ./configure -device linux-generic-g++ -device-option CROSS_COMPILE=mipsel-linux-gnu-

include(../common/linux_device_pre.conf)
include(../common/linux_device_post.conf)
load(qt_config)
QMAKE_LFLAGS=
QMAKE_CFLAGS=
QMAKE_CXXFLAGS=
```

I don't know if this is expected, what this depends on, and why we only want to propagate these flags in this way when cross-compiling.

amjoseph-nixpkgs replied:
You're absolutely right, those lines are unnecessary.  Will fix, and then squash.

amjoseph-nixpkgs replied:
Fix is in cf5fef3fac25d3cbc3ae6de53211d21eb0e631eb.

Will squash once I rebuild to verify.

amjoseph-nixpkgs replied:
Works, squashed.

OPNA2608 commented on pkgs/development/libraries/qt-5/modules/qtbase.nix:
```diff
@@ -161,6 +171,13 @@ stdenv.mkDerivation (finalAttrs: {
     export MAKEFLAGS+=" -j$NIX_BUILD_CORES"
 
     ./bin/syncqt.pl -version $version
+  '' + lib.optionalString (stdenv.buildPlatform.canExecute != stdenv.hostPlatform) ''
```
```suggestion
  '' + lib.optionalString (stdenv.buildPlatform != stdenv.hostPlatform) ''
```

Commit:
qt5.qtbase: implement @OPNA2608 suggestion

From here: https://github.com/NixOS/nixpkgs/pull/227900#discussion_r1299351842

amjoseph-nixpkgs commented:
@OPNA2608 let me know what you think of the latest push.

OPNA2608 commented on pkgs/development/libraries/qt-5/modules/qtbase.nix:
```diff
@@ -161,6 +171,11 @@ stdenv.mkDerivation (finalAttrs: {
+    # QT's configure script will refuse to use pkg-config unless these two environment variables are set
+    export PKG_CONFIG_SYSROOT_DIR=/
+    export PKG_CONFIG_LIBDIR=${lib.getLib pkg-config}/lib
+    echo "QMAKE_PKG_CONFIG=''$''${CROSS_COMPILE}pkg-config" >> mkspecs/devices/${qtPlatformCross stdenv.hostPlatform}/qmake.conf
```
Built the change. This gives me a broken result in the final qmake.conf:

```
QMAKE_PKG_CONFIG=1{CROSS_COMPILE}pkg-config
```

because this is quoted with `"` and [the shell expands the `$$` to its pid](https://www.gnu.org/software/bash/manual/html_node/Special-Parameters.html#index-_0024) during the echoing.

```suggestion
    echo 'QMAKE_PKG_CONFIG=''$''${CROSS_COMPILE}pkg-config' >> mkspecs/devices/${qtPlatformCross stdenv.hostPlatform}/qmake.conf
```

This gives me the correct result:

```
QMAKE_PKG_CONFIG=$${CROSS_COMPILE}pkg-config
```

amjoseph-nixpkgs replied:
I hate bash.

OPNA2608 reviewed (changes requested):
And I feel like the modifying of that config file really belongs into `postPatch` more than `preConfigure`.

Commit:
Update pkgs/development/libraries/qt-5/modules/qtbase.nix

Co-authored-by: Christoph Neidahl <christoph.neidahl@gmail.com>

OPNA2608 commented:
Sorry to nit, but the title of the last commit doesn't match the commit conventions. Prolly fine to just squash it into the preceding one.

amjoseph-nixpkgs commented:
So apparently during the ~four months it took for this to be reviewed, somebody went and simply started deleting qt5 support from a bunch of packages.  And now they're claiming those deletions can't be reverted because it might cause some users' browser cookies to be deleted *after they are warned and given an opportunity to not delete their cookies*.

This is incredibly demoralizing.

I think nixpkgs has a real organizational/cultural problem; the swarms of drive-by version-bump committers massively outnumber the people willing to put in long sustained effort on the hard underlying plumbing of nixpkgs.  The latter have PRs that take way longer to review, and have to suffer way more merge conflicts in the meantime.

At some point we're going to lose the plumbers.  And all that will be left will be the yolo-version-bump committers.  That's not going to be a healthy software distribution.

dotlambda commented:
Shouldn't this have gone to staging?

dotlambda commented:
> And now they're claiming those deletions can't be reverted because it might cause some users' browser cookies to be deleted _after they are warned and given an opportunity to not delete their cookies_.

False. A revert should be discussed before it's done, that's all.

dotlambda commented:
Here are my deleted comments:

Shouldn't this have gone to staging?

> And now they're claiming those deletions can't be reverted because it might cause some users' browser cookies to be deleted after they are warned and given an opportunity to not delete their cookies.

False. A revert should be discussed before it's done, that's all.

samueldr commented:
@amjoseph-nixpkgs can you explain why comments ended-up being deleted here?

amjoseph-nixpkgs commented:
Sorry, I should have marked them off-topic, which they are.

samueldr commented:
No.

samueldr commented:
***This should have gone to staging.***

This is not the appropriate behaviour here.

You can't just decide to flip up the game board when you're mad about how things are going in the game.

We have processes for development. Mass rebuilds ***cannot*** go in master. That's always been the case.

samueldr commented on pkgs/development/libraries/qt-5/qtModule.nix:
```diff
@@ -74,4 +79,7 @@ mkDerivation (args // {
     platforms = platforms.unix;
   } // (args.meta or {});
+
+} // lib.optionalAttrs (stdenv.hostPlatform != stdenv.buildPlatform) {
+  dontWorryAboutQtMismatch = true;
```
(Now, since I've been made aware of this PR... let's look into it...)

Is this an appropriate change? Mismatches in Qt will lead to severe issues down the line.

I suspect it's a cross-compilation idiosyncrasy here, but without a comment explaining it, I have no way to know for sure.

If a platform name ended-up breaking this check, the platform name should be added to the check.

vcunat commented:
Another issue with this PR is that it breaks build of qtbase on x86_64-darwin &ndash; and you could've seen the error on ofBorg already.  Hydra: https://hydra.nixos.org/build/232824447/nixlog/2/tail

RaitoBezarius commented:
I will address this publicly.

> So apparently during the ~four months it took for this to be reviewed, somebody went and simply started deleting qt5 support from a bunch of packages. And now they're claiming those deletions can't be reverted because it might cause some users' browser cookies to be deleted _after they are warned and given an opportunity to not delete their cookies_.

Pull request discovery or "who is working on that" discovery is a difficult problem and I acknowledge this is a very frustrating experience.

I am not a domain expert in those areas and I cannot judge for the claim on impossibility of revert the deletions, I know that browser profiles have a migration system, and they rarely expect to go in both directions. Therefore, the claim is _at least_ realistic. As we err on the caution, and we act generally in favor of the largest numbers (yes, this sucks but more on that later.).

The call of reviving a new package set for qt5 or qt6 is a valid option and has been proposed, I know that in all of that frustration, I can understand people losing their cool and moving on actions.

> This is incredibly demoralizing.

I am sorry that you feel like this, all I can say is that nixpkgs is an inherently frustrating project, we all work to reduce it in the long run by building better tools for contributors themselves, improving our CI, improving our tests, improving a lot of things, it's not a work that receives a lot of attention.

I would love to have nixpkgs be the perfect place for people having exotic setups on a daily basis but we are already struggling with tier 2 platforms, it's not realistic to push the burden on people regarding cross compilation yet.

All of this should be the best effort basis and engaging in good faith.

> I think nixpkgs has a real organizational/cultural problem; the swarms of drive-by version-bump committers massively outnumber the people willing to put in long sustained effort on the hard underlying plumbing of nixpkgs. The latter have PRs that take way longer to review, with far fewer qualified reviewers, and they have to suffer way more merge conflicts in the meantime.

I would argue this is a labor resource problem rather than organizational/cultural, as you can see, nixpkgs is decentralized in their collaboration, you do not bother joining the Matrix platform to coordinate with anyone and this is mostly fine except in moment of higher stress and nervousness that cannot be handled calmly by GitHub discussions.

People often spend time posting about their avenues, e.g. removing Qt 5, etc. They ask, re-ask, ask for reviews, etc., and the model works to a certain extent for a lot of contributors.

On many instances, I have seen / reviewed your work on the sole basis that I care about your work and I enjoy it.

Overall, I feel like you have a very special problem that is shared by everyone who is doing their collaboration without interacting on the RFC-mandated (which, for public record, is driving me completely mad on a daily basis.) collaboration platform. I conjure you: please solve this issue moving forward. If I am not wrong, I believe this is one of the biggest driver for your demoralization and inability to solve this type of incident in a polite way off-GitHub.

> At some point we're going to lose the plumbers. And all that will be left will be the yolo-version-bump committers. That's not going to be a healthy software distribution.

We are surely going to lose the plumbers for some areas, but not everything. It's fine if nixpkgs does not improve its plumbers for a while, it's a shame though because you are doing awesome and talented work in cross compilation and GCC space.

But, as a release manager, I cannot stand up for someone who caused such an incident, as you know, this is not the first friction you created.

Just to be clear and correct me if I'm wrong:

- this PR was sent in reaction to the other PR on qutebrowser
- this PR caused breakage for Darwin as @vcunat mentions it: it could have been caught using OfBorg
- this PR caused mass rebuilds: build infra confirms it even though you disagreed or ignored
- in this PR, you removed comments you considered as off-topic: this is a dangerous violation of privileges I believe is a **total nogo** from a nixpkgs **committer**
- in the other PR, you engaged into a confrontational revert which was reverted again: such situations should always be solved politely or at least escalated to an intermediate of authority of nixpkgs, e.g. GitHub organization owners, release managers, etc. In theory, maintainers of said packages should have the last say in conflict resolution except in presence of special cases: release management concerns, security team concerns, etc.
- As you know, nixpkgs is a public good and its resources are important, we cannot aim for perfect support for any platforms, cross is bound to be a non-concern from a lot of maintainers (which I *also* regret FWIW.) because it's unrealistic to expect from them to handle it. Therefore, your *entitlement* for supporting your usecases on a short notice is uncalled-for — I believe you have the skills to maintain patches on the top of your nixpkgs HEADs.

I am sorry if this sounds harsh, and I am willing to apologize for the insufficiencies of nixpkgs as I am probably responsible for not doing enough for this project. But I am not okay with antagonizing contributors, and we are not at the first instance of collaboration events charged with friction/confrontational remarks/etc. from your end.

I conjure you to change your ways. 

We are all in the same boat, including the "yolo-version-bump committers" who I would urge you to find more respectful words to describe them because they are doing the hard work of testing for the tier1/tier2 platforms we do have and maintaining this software distribution up-to-date.

Once we get there, I cannot guarantee that collaboration will be frustrationless, the people doing the plumbing work always had it hard in nixpkgs and this is merciless work, but expectation-wise, we will definitely be able to make a dent and move the agenda forward and focus on more and more interesting stuff.

zimbatm commented:
I share Amjoseph's frustration; people who do large refactors are not well accompanied.

The longer the PR is dragging along, the more people will participate in it, with potentially conflicting opinions. It feels like trying to move with more and more weights attached to the back. With no end in sight (insert dramatic tone)! And because we are spread thin, the quality of the reviews isn't always there either.

Potentially a better system would quickly identify and triage those PRs. Decide if fundamentally it's right, and if yes, assign a fixed review buddy who is in the same camp and helps move things along.
